src/kvstore.c

```diff
@@ -568,14 +577,23 @@ kvstoreIterator *kvstoreIteratorInit(kvstore *kvs) {
 void kvstoreIteratorRelease(kvstoreIterator *kvs_it) {
     dictIterator *iter = &kvs_it->di;
     dictResetIterator(iter);
-
+    freeDictIfNeeded(kvs_it->kvs, kvs_it->didx);
     zfree(kvs_it);
 }
 
 /* Returns next dictionary from the iterator, or NULL if iteration is complete. */
 dict *kvstoreIteratorNextDict(kvstoreIterator *kvs_it) {
     if (kvs_it->next_didx == -1)
         return NULL;
+
+    /* The dict may be deleted during the iteration process, so here need to check for NULL. */
+    if (kvs_it->didx != -1 && kvstoreGetDict(kvs_it->kvs, kvs_it->didx)) {
+        /* Before we move to the next dict, reset the iter of the previous dict. */
+        dictIterator *iter = &kvs_it->di;
+        dictResetIterator(iter);
+        freeDictIfNeeded(kvs_it->kvs, kvs_it->didx);
+    }
```

Comment:
was that (dictResetIterator) missing regardless of the main topic of this PR?

Reply:
yes

Reply:
btw, we currently have no scenario that elements will be deleted in kvstoreIterator process. I discovered it during the fix (i think we also need to deal with it together)

Reply:
to avoid confusion, let's mention that unrelated change in the top comment.